### PR TITLE
[WIP] Goodbye jemalloc

### PR DESCRIPTION
**Pull request by japaric**
Out of free version (for testing purposes): https://github.com/japaric/alloc-jemalloc

Minimally tested on x86_64 Linux. Still haven't fixed/updated the in-tree tests

cc @alexcrichton @brson

### CODE REVIEW

File: src/librustc_back/target/apple_base.rs [hunk 1]
```diff
@@ -44,7 +44,7 @@ pub fn opts() -> TargetOptions {
         dll_suffix: ".dylib".to_string(),
         archive_format: "bsd".to_string(),
         pre_link_args: Vec::new(),
-        exe_allocation_crate: super::maybe_jemalloc(),
+        exe_allocation_crate: "alloc_system".to_owned(),
```
Comment (japaric) [inline src/librustc_back/target/apple_base.rs]: Hmm, I wonder if we can remove this now.

Comment (japaric) [inline src/librustc_back/target/apple_base.rs]: (the `exe_allocation_crate` field and logic related to it)

**Comment by eddyb** [comment]
r? @alexcrichton 

cc @Mark-Simulacrum @nikomatsakis @nnethercote on rustc perf

**Comment by steveklabnik** [comment]
I had asked @alexcrichton and @brson about this the other day; they said we couldn't do this before we made the interface stable. Maybe this will change their minds?

**Comment by sfackler** [comment]
The interface stabilization issue is still there: https://github.com/japaric/alloc-jemalloc/blob/master/src/lib.rs#L13

**Comment by alexcrichton** [comment]
Thanks for the PR @japaric! I definitely sympathize with how much of a maintenance burden jemalloc is at this time.

Unfortunately though I don't think we're in a position to accept this PR at this time. As you've found out it's technically quite easy to remove jemalloc, but as a stable feature of Rust today this would be a breaking change with an unacceptable path forward I think.

We've had jemalloc as the default allocator since 1.0, and it objectively performs better than any other allocator in a variety of workloads. Programs may or may not be relying on this performance. I know that the compiler, at least, has historically performed better under jemalloc than the system allocator.

If we were to remove jemalloc from in tree then there is no transition path for stable users using jemalloc. They'd all be forced to switch to nightly to regain perf losses, which would be an unacceptable transition in my opinion.

Dealing with jemalloc is a very high priority amongst the libs team. It's slated to be the next point of discussion after we've sorted through the stories of scenarios and SIMD.

**Comment by arthurprs** [comment]
> If we were to remove jemalloc from in tree then there is no transition path for stable users using jemalloc. They'd all be forced to switch to nightly to regain perf losses, which would be an unacceptable transition in my opinion.

:+1:

**Comment by japaric** [comment]
@alexcrichton Fair enough. Closing until the libs team gets to talk about it.

**Pull request closed without merging**